src/pal/pal_device.cpp

```diff
@@ -0,0 +1,541 @@
+            break;
+    }
+
+    Pal::QueueCreateInfo queue_create_info = {};
+    queue_create_info.queueType = queue_type;
+    queue_create_info.engineType = engine_type;
+
+    Pal::Result result = Pal::Result::ErrorUnknown;
+
+    // Create the queue
+    const size_t queue_size = device_->GetQueueSize(queue_create_info, &result);
+    CHECK_PAL(result, "GetQueueSize()");
+
+    result = device_->CreateQueue(queue_create_info, malloc(queue_size), &queue);
+    CHECK_PAL(result, "CreateQueue()");
+
+    // Create the command buffer.
+    Pal::CmdBufferCreateInfo cmd_buffer_create_info = {};
+    cmd_buffer_create_info.pCmdAllocator = cmd_allocator_;
+    cmd_buffer_create_info.queueType = queue_type;
+    cmd_buffer_create_info.engineType = engine_type;
+
+    const size_t cmdBufSize = device_->GetCmdBufferSize(cmd_buffer_create_info, &result);
+    CHECK_PAL(result, "GetCmdBufferSize()");
+
+    return device_->CreateCmdBuffer(cmd_buffer_create_info, malloc(cmdBufSize), &cmd_buffer);
+}
+
+PalDevice::GpuVirtAddr_t PalDevice::allocate_gpu_memory(Pal::gpusize size_in_bytes, Pal::GpuHeap heap) {
+    Pal::IGpuMemory* memory;
+    Pal::Result result = allocate_memory(size_in_bytes, heap, &memory);
+    CHECK_PAL(result, "allocate_gpu_memory(GpuHeapInvisible)");
+    const auto gpu_virtual_addr = track_memory(memory);
+    return gpu_virtual_addr;
+}
+
+Pal::Result PalDevice::allocate_memory(
+    Pal::gpusize size_in_bytes, Pal::GpuHeap heap, Pal::IGpuMemory** gpu_memory_pp, Pal::gpusize alignment) {
+    Pal::GpuMemoryHeapProperties heap_properties[Pal::GpuHeapCount] = {};
+    Pal::Result result = device_->GetGpuMemoryHeapProperties(&heap_properties[0]);
+    CHECK_PAL(result, "GetGpuMemoryHeapProperties()");
+
+    Pal::GpuMemoryCreateInfo create_info = {};
+    create_info.alignment = alignment;
+    create_info.size = size_in_bytes;
+    create_info.flags.cpuInvisible = ((heap_properties[heap].flags.cpuVisible == 0) ? 1 : 0);
+    create_info.vaRange = Pal::VaRange::Default;
+    create_info.heapCount = 1;
+    create_info.priority = Pal::GpuMemPriority::Normal;
+    create_info.priorityOffset = Pal::GpuMemPriorityOffset::Offset0;
+    create_info.heaps[0] = heap;
+    create_info.mallPolicy = Pal::GpuMemMallPolicy::Always;
+
+    const size_t gpu_memory_size = device_->GetGpuMemorySize(create_info, &result);
+    CHECK_PAL(result, "GetGpuMemorySize()");
+
+    result = device_->CreateGpuMemory(create_info, malloc(gpu_memory_size), gpu_memory_pp);
+    CHECK_PAL(result, "CreateGpuMemory()");
+
+    Pal::GpuMemoryRef mem_ref = {};
+    mem_ref.pGpuMemory = *gpu_memory_pp;
+
+    // Make this allocation permanently resident. We might want to track  which allocations are actually being
+    // used and only make those resident but in AnyDSL usually the user takes care of releasing memory as soon
+    // as it's not used anymore.
+    return device_->AddGpuMemoryReferences(1, &mem_ref, nullptr, Pal::GpuMemoryRefCantTrim);
```

Comment:
Tracking allocations like that is probably unworkable since there's no concept of binding in AnyDSL's runtime API, we do everything with buffer addresses or handles that we can provide through arbitrary channels.

Yes - the CL backend will bind buffers that are passed to the main function, but that's because old-school CL stinks ! I'm fairly sure we don't want that going forward.

If GPU memory pressure proves a real issue we can always expose the concept of residency.

Reply:
Alright :+1: I'll adjust the comment to only state that the allocation is marked as resident until it's released explicitly by the user program.